### PR TITLE
collecting GCP clusters data

### PR DESCRIPTION
**Pull request by AnnuCode**
## Problem

Collect GCP Container clusters data

## Solution

Created `container` package to collect GCP container clusters data.

## Changes Made

- created `container` package.
- transferred the `ExtractZoneFromURL` function to `utils` package in `region.go` file as it is also used in the compute package. 

## Notes

- `Link` of the GKE needs to be added in the `models.Resource` slice.
- `Nodeconfig` is used which is deprecated according to the docs. It can be replaced using `Nodepool`.

## Checklist

- [x] Code follows the <a href="https://github.com/tailwarden/komiser/blob/master/CONTRIBUTING.md">contributing</a> guidelines
- [x] Any dependencies have been added to the project, if necessary

## Reviewers

@eneskaya

### CODE REVIEW

**Review by mlabouardy — COMMENTED** [review]
@AnnuCode do you need some help for the PR to be ready for review? :)

**Comment by AnnuCode** [comment]
@mlabouardy Hi, `Link` of the `GKE` needs to be added in the `models.Resource` slice but I do not know the correct format of the link.

File: providers/gcp/container/container.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+		resources = append(resources, models.Resource{
+			Provider:   "GCP",
+			Account:    client.Name,
+			Service:    "Cluster",
+			ResourceId: cluster.GetId(),
+			Region:     zone,
+			Name:       cluster.GetName(),
+			FetchedAt:  time.Now(),
+			Tags:       tags,
+			//Link
```
Comment (eneskaya) [inline providers/gcp/container/container.go:57]: Hey @AnnuCode, so for the link, this should do:

```
https://console.cloud.google.com/kubernetes/clusters/details/[REGION]/[CLUSTER_NAME]/details?project=[PROJECT_ID]
```

You should have all the info that is needed to fill it in

File: providers/gcp/container/container.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+func Clusters(ctx context.Context, client providers.ProviderClient) ([]models.Resource, error) {
+	resources := make([]models.Resource, 0)
+	clusterClient, err := container.NewClusterManagerClient(ctx, option.WithCredentials(client.GCPClient.Credentials))
+	if err != nil {
+		logrus.WithError(err).Errorf("failed to create cluster client")
+		return resources, err
+	}
+
+	req := &containerpb.ListClustersRequest{
+
```
Comment (eneskaya) [inline providers/gcp/container/container.go:26]: Uneeded new line here

File: utils/regions.go [hunk 3]
```diff
@@ -954,4 +954,8 @@ func NormalizeRegionName(regionName string) string {
 	lowercased := strings.ToLower(regionName)
 	dashReplaced := strings.Replace(lowercased, "_", "-", -1)
 	return dashReplaced
-}
\ No newline at end of file
+}
+
+func ExtractZoneFromURL(url string) string {
```
Comment (eneskaya) [inline utils/regions.go:959]: As this function is very specific to GCP, please rename to `GcpExtractZoneFromURL` or add a comment describing this please

File: providers/gcp/container/container.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	}
+	clusters, err := clusterClient.ListClusters(ctx, req)
+	if err != nil {
+		logrus.WithError(err).Errorf("failed to collect clusters")
+		return resources, err
+	}
+
+	for _, cluster := range clusters.Clusters {
+		tags := make([]models.Tag, 0)
+		//according to docs, NodeConfig is deprecated and is advised to use node_pool.config
```
Comment (eneskaya) [inline providers/gcp/container/container.go:37]: one more remark, if it's deprecated to use NodeConfig, can you change this to the non-deprecated way?

Comment (AnnuCode) [reply providers/gcp/container/container.go:37]: @eneskaya yes sure, working on it.

**Comment by eneskaya** [comment]
Hey @AnnuCode I wanted to test your code, but it seems there is a critical bug which leads to komiser crashing. Please see screenshot. You can ping me once the issue is resolved and I can have a look again.

Screenshot 2023-04-06 at 08 56 33

**Comment by AnnuCode** [comment]
HI @mlabouardy , I think the program was crashing(invalid memory address) because I forgot to write the `return` statement in the error check associated with `NodePool` which is being shown in the attached photo above('failed to collect labels'). The issue is after I have written the return statement in the error check, I am not able to check if its the right fix as I am getting 'Network Request Error' after running `go build` and `./komiser start` on my `feature/666` branch. Is it because of the lack of cloud credentials in my case? The program is not crashing anymore with the above shown error but maybe I am not able to reproduce it properly?

**Comment by mlabouardy** [comment]
> HI @mlabouardy , I think the program was crashing(invalid memory address) because I forgot to write the `return` statement in the error check associated with `NodePool` which is being shown in the attached photo above('failed to collect labels'). The issue is after I have written the return statement in the error check, I am not able to check if its the right fix as I am getting 'Network Request Error' after running `go build` and `./komiser start` on my `feature/666` branch. Is it because of the lack of cloud credentials in my case? The program is not crashing anymore with the above shown error but maybe I am not able to reproduce it properly?

the 'Network Request Error' error was fixed last week, if you rebase your branch with `develop` (make sure you pull latest changes from develop first) you won't have the issue anymore :)

**Comment by AnnuCode** [comment]
@mlabouardy Tried but could not do it. I started with `git fetch upstream` --> `git merge upstream/develop`  from the feature branch --> `git rebase upstream/develop` from the feature branch. After resolving conflicts and doing `git rebase --continue` VSCode says failed to execute git. Also created a mess in this branch in the process.

**Comment by AnnuCode** [comment]
@mlabouardy @eneskaya  After resolving conflicts in the rebase process, this message is there. Can you please have a look? 

git_rebase

**Review by mlabouardy — APPROVED** [review]
@AnnuCode I've fixed the missing dependencies, thanks

**Pull request merged**